### ISSUE TITLE
Inappropriate HTTP status codes in htsget

### ISSUE THREAD

**Issue by jeromekelleher** [samtools/hts-specs#186]
The [htsget spec](http://samtools.github.io/hts-specs/htsget.html) currently requires that the  server should return a HTTP status 409 when the client requests a format that the server doesn't support (i.e. ``format=something_weird``). [HTTP 409](https://httpstatuses.com/409) indicates that "the request could not be completed due to a conflict with the current state of the target resource", which seems like quite a stretch from "you've requested a format I don't support".  

While being concrete about the HTTP status codes is a great idea, I think it's probably a bad idea to co-opt these (I would have thought) obscure error codes for our own purposes for two reasons:

1. There will be clients that ignore the JSON payload returned in error conditions, and just raise the HTTP error directly. In this case, the user will be presented with very confusing errors. For example, using ``requests`` if I make a request for a bad file type I'll get back the error "409 Client Error: Conflict for url: ".  Googling for this sends me off in completely the wrong direction. A simple "400 Bad Request" would be much better in this case I think, because even if it is vague at least it's not falsely precise.

2. We should assume that there will be a tall stack of other HTTP processing entities between the client and the server. These will interpret the HTTP error codes according to the original specifications (which are quite precise about what clients should and should not do in the context of the various error codes), and may take actions based on them. Co-opting the error codes for our own specific needs could lead to very weird and difficult to debug behaviour.

So, I would argue that we should limit ourselves to a small number of widely used and well understood HTTP status codes (400, 403, 404, 500?). Certainly we should have some mechanism for passing back more specific information to the client, but this should be held in the JSON payload, not the HTTP status.

Thoughts? cc'ing @mlin and @dglazer here, as I'm not sure if they are watching this repo.

**Comment by vadimzalunin** [comment]
I think this error is basically saying that the data does exists but in some other format(s). Clients may want to recover after this error unlike say "400 Bad Request".

**Comment by jeromekelleher** [comment]
> Clients may want to recover after this error unlike say "400 Bad Request".

I totally agree, but I think we need to make a distinction between the HTTP client and the htsget client. Currently, for the htsget server to communicate with the htsget client, we are giving the HTTP client misleading information (HTTP 409, in this example).  Htsget is layered on top of HTTP, and by co-opting HTTP error codes for our own specific purposes we're breaking this layering. To preserve the layering, the error information specific to the htsget protocol should be contained in the JSON body of the response.

So, to be clear, in this example I think we should return HTTP 400 (Bad Request) with a body like

```json
{ 
    "type": "UnsupportedFormat", 
     "message": "The requested file format is not supported by the server"
}
```

The htsget client can parse the JSON, determine the type of error and then act on this information as it sees fit.

**Comment by vadimzalunin** [comment]
from https://httpstatuses.com/409:

> 409 CONFLICT
> The request could not be completed due to a conflict with the current state of the target resource. This code is used in situations where the user might be able to resolve the conflict and resubmit the request.
> 

Not sure if the current spec breaks any HTTP compatibility by hijacking this code. Will implementations ever want to use 409 to communicate some other meaning?

Imo HTTP codes are convenient to report a handful of errors. I'm not strongly opposed to the suggestion but I don't see a strong reason for the change either unless there is more errors to report.

**Comment by jeromekelleher** [comment]
>  Will implementations ever want to use 409 to communicate some other meaning?

I think we need to consider here not just htsget servers and clients, but whole HTTP infrastructure that surrounds them. For example, the htsget server might run on a server like uwsgi, which is reverse proxied by nginx for within-host load balancing (perhaps with a memcached thrown in), which is in turn run by Amazon elastic load balancer for fault tolerance and cross host load balancing and scaling. I don't know if any of these would ever use a 409 for their own purposes, but I think it's a lot safer to not take the risk. 

> Imo HTTP codes are convenient to report a handful of errors. I'm not strongly opposed to the suggestion but I don't see a strong reason for the change either unless there is more errors to report.

I don't know how many error codes we are going to need, but it seems sensible to leave the door open for extensibility in case we do need more.

**Comment by vadimzalunin** [comment]
> I don't know if any of these would ever use a 409 for their own purposes, but I think it's a lot safer to not take the risk.
I don't see why they can't use 400 as well.

**Comment by dglazer** [comment]
Re 409 in particular -- I agree, it doesn't sound appropriate for our use.
The definition goes on to say that "This code is only allowed in situations
where it is expected that the user might be able to resolve the conflict
and resubmit the request.", and that "Conflicts are most likely to occur in
response to a PUT request.", neither of which fit here.

Re 400 vs. 4xx -- I don't know modern best practices. I agree with Jerome
that many many servers return 400 and 404, and that therefore it's safe to
assume that infrastructure stacks are good at passing them through. I
believe that the same *should* be true for other 4xx codes, including 409,
but since it's rarely used, it might not be safe to count on it.

Bottom line -- I agree with Jerome, 400 makes more sense than 409 here.
Copied Mike Lin, since he may have some history behind the original
proposal.

**Comment by mlin** [comment]
Uncharacteristically, I don't have too strong of an opinion here. This webpage seems to agree with Jerome's suggested approach.
https://apigee.com/about/blog/technology/restful-api-design-what-about-errors

Get Outlook for Android<https://aka.ms/ghei36>


________________________________

**Comment by vadimzalunin** [comment]
Given the consensus, @jeromekelleher - feel free to submit a PR.

**Comment by jmarshall** [comment]
Completely agree that this use of 409 is not in line with 409's defined meaning.  It is in line with the meaning of 406, so we might consider either that or the generic 400.

[Separately…]
> Similarly they may use 400 as well.

400's meaning is _generic-client-error_, so this use of it would be in accordance with its meaning.
